exchanges/orderbook/depth.go

```diff
@@ -0,0 +1,352 @@
+func (d *Depth) IsRestSnapshot() bool {
+	d.m.Lock()
+	defer d.m.Unlock()
+	return d.restSnapshot
+}
+
+// LastUpdateID returns the last Update ID
+func (d *Depth) LastUpdateID() int64 {
+	d.m.Lock()
+	defer d.m.Unlock()
+	return d.lastUpdateID
+}
+
+// IsFundingRate returns if the depth is a funding rate
+func (d *Depth) IsFundingRate() bool {
+	d.m.Lock()
+	defer d.m.Unlock()
+	return d.isFundingRate
+}
+
+// Alert defines fields required to alert sub-systems of a change of state to
+// re-check depth list
+type Alert struct {
+	// Channel to wait for an alert on.
+	forAlert chan struct{}
+	// Lets the updater functions know if there are any routines waiting for an
+	// alert.
+	sema uint32
+	// After closing the forAlert channel this will notify when all the routines
+	// that have waited, have either checked the orderbook depth or finished.
+	wg sync.WaitGroup
+	// Segregated lock only for waiting routines, so as this does not interfere
+	// with the main depth lock, acts as a rolling gate.
+	m sync.Mutex
+}
+
+// alert establishes a state change on the orderbook depth.
+func (a *Alert) alert() {
+	// CompareAndSwap is used to swap from 1 -> 2 so we don't keep actuating
+	// the opposing compare and swap in method wait. This function can return
+	// freely when an alert operation is in process.
+	if !atomic.CompareAndSwapUint32(&a.sema, 1, 2) {
+		// Return if no waiting routines or currently alerting.
+		return
+	}
+
+	go func() {
+		// Actuate lock in a different routine, as alerting is a second order
+		// priority compared to updating and releasing calling routine.
+		a.m.Lock()
+		// Closing; alerts many waiting routines.
+		close(a.forAlert)
+		// Wait for waiting routines to receive alert and return.
+		a.wg.Wait()
+		atomic.SwapUint32(&a.sema, 0) // Swap back to neutral state.
+		a.m.Unlock()
+	}()
+}
+
+// Wait pauses calling routine until depth change has been established via depth
+// method alert. Kick allows for cancellation of waiting or when the caller has
+// has been shut down, if this is not needed it can be set to nil. This
+// returns a channel so strategies can cleanly wait on a select statement case.
+func (a *Alert) Wait(kick <-chan struct{}) <-chan bool {
+	reply := make(chan bool)
+	a.m.Lock()
+	a.wg.Add(1)
+	if atomic.CompareAndSwapUint32(&a.sema, 0, 1) {
+		a.forAlert = make(chan struct{})
+	}
+	go a.hold(reply, kick)
+	a.m.Unlock()
+	return reply
+}
+
+// hold waits on either channel in the event that the routine has finished or an
+// alert from a depth update has occurred.
+func (a *Alert) hold(ch chan<- bool, kick <-chan struct{}) {
+	select {
+	// In a select statement, if by chance there is no receiver or its late,
+	// we can still close and return, limiting dead-lock potential.
+	case <-a.forAlert: // Main waiting channel from alert
+		select {
+		case ch <- false:
+		default:
+		}
+	case <-kick: // This can be nil.
+		select {
+		case ch <- true:
+		default:
+		}
+	}
+	a.wg.Done()
```

Comment:
Last question I think! Should this be below close?

Reply:
I put this before close because this directly relates to the alert routine's a.wg.Wait(). It has a higher priority because it wants to swap back to a neutral state and unlock, if it stays locked the biggest impedance on the alert system is the Wait() call because there can be updates occurring while the alert system is in this locked state (potentially) and at this point, there could be routines blocked by the mutex slip ring functionality and not actually get alerted (small chance). Also, an alert has happened either by forAlert or kick, the close channel just makes sure if the calling routine is late it will still fire in this context until a new Wait() call.
That was my intention anyway, if logic is incorrect let me know.

Reply:
Yeah it doesn't look like it causes much impact, so its all g 👍